### ISSUE TITLE
Future.wait casting is not solved by Records

### ISSUE THREAD

**Issue by srawlins** [dart-lang/language#2303]
One of the five paragraphs under Motivation is about `Future.wait`:

https://github.com/dart-lang/language/blob/df9574472a7a84aa518b88794c5ec9ff6349f170/working/0546-patterns/records-feature-specification.md

> You've probably noticed this if you've ever used `Future.wait()` to wait on a
couple of futures of different types. You end up having to cast the results back
out since the type system no longer knows which element in the returned list has
which type.

As far as I can tell, Records cannot solve this for the existing generic `Future.wait`. There is not a new signature we can use for `Future.wait` that would return a record, for example. (You could define a nice `Future<(T0, T1)> wait2(Future<T0> f0, Future<T1> f1)`, `wait3`, `wait4`, etc. though.)

It looks to me like this is more directly solved with the [cast binder](https://github.com/dart-lang/language/blob/master/working/0546-patterns/patterns-feature-specification.md#cast-binder) pattern:

```dart
var [a as int, b as String] = await Future.wait([aFuture, bFuture]);
```

Not sure what the AI is here; perhaps the motivating example could be moved to the "cast binder" pattern section...

**Comment by munificent** [comment]
> we can safely say that `await record` will effectively pointwisely await _every component_ of the record, and evaluate to the result of awaiting a `Future<recordType'>` with the same structure as the original, and element types being the _flatten_ of the original element types.

Can we also do that for lists too, so that awaiting a list of futures yields a future of a list of the results? Sometimes you do need to wait for a dynamically-sized homogeneous list of asynchronous tasks. (I suppose the answer is "no" because there could be some user-defined class that implements *both* `List` and `Future` :( ).

Yes, I'm on board with this. Maybe we should carve it out as a separate proposal since the main patterns one already has a lot of surface area?

**Comment by munificent** [comment]
Barring that, you could do something fairly nice with extensions:

```dart
extension FutureRecord2Extension<T1, T2> on (Future<T1>, Future<T2>) {
  Future<(T1, T2)> wait() async {
    var (f1, f2) = this;
    var [e1 as T1, e2 as T2] = await Future.wait([f1, f2]);
    return (e1, e2);
  }
}

extension FutureRecord3Extension<T1, T2, T3> on (Future<T1>, Future<T2>, Future<T3>) {
  Future<(T1, T2, T3)> wait() async {
    var (f1, f2, f3) = this;
    var [e1 as T1, e2 as T2, e3 as T3] = await Future.wait([f1, f2, f3]);
    return (e1, e2, e3);
  }
}

// Other arities...
Future<int> asyncThing() ...
Future<String> another() ...

main() async {
  var (n, s) = await (asyncThing(), another()).wait();
}
```

**Comment by lrhn** [comment]
Like the extensions, hate the dynamic typing being necessary. Also doesn't scale well with named record elements.

A more direct approach, without using `Future.wait`, would be:

```dart
extension FR2X<T1, T2> on (Future<T1>, Future<T2>) {
  Future<(T1, T2)> wait() async {
    Completer<(T1, T2)> c = Completer.sync();
    T1? v1;
    T2? v2;
    var count = 0;
    List<AsyncError>? errors;
    void tryComplete() {
       if (++count == 2) {
         if (errors == null) {
           c.complete((v1 as T1, v2 as T2));
         } else {
           c.completeError(AwaitRecordException<(T1?, T2?)>((v1, v2), errors));
         }
       }
    }
    var (f1, f2) = this;
    unawaited(f1.then((v) { v1 = v; tryComplete()}, onError: (Object e, StackTrace s) {
      (errors ??= []).add(AsyncError(e, s); tryComplete(); 
    }));
    unawaited(f2.then((v) { v2 = v; tryComplete()}, onError: (Object e, StackTrace s) {
      (errors ??= []).add(AsyncError(e, s); tryComplete(); 
    }));
    return c.future;
  }
}
```
Doesn't scale to other record structures any better. This really begs to be a language or library feature, and for best typing, it really should be a language feature.

**Comment by leafpetersen** [comment]
> Yes, I'm on board with this. Maybe we should carve it out as a separate proposal since the main patterns one already has a lot of surface area?

I'd like to re-surface this.  I think this proposal has a lot of potential value, and I think if we're going to do it, we need to do it now (since once the first record gets awaited, this becomes a breaking change).

Thoughts on this?

cc @munificent @lrhn @eernstg @jakemac53 @natebosch @stereotype441

**Comment by leafpetersen** [comment]
I see that @lrhn has a proposal for this [here](https://github.com/dart-lang/language/issues/2321).

**Comment by munificent** [comment]
Closing this as a duplicate of #2321 since that has covers the same territory and has more details.

**Comment by srawlins** [comment]
Thanks for all of the discussion!
